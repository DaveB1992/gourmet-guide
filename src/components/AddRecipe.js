import React, { useState } from "react";
import axios from "axios";
import "../styles/addrecipe.css";
import Alert from "./Alert";

const AddRecipe = () => {
  const initialState = {
    fields: {
      title: "",
      cuisine: "",
      dietaryRequirements: "",
      ingredients: [{ name: "", measurement: { value: "", metric: "" } }],
      instructions: "",
      prepTime: "",
      cookingTime: "",
      servings: "",
    },
    alert: {
      message: "",
      isSuccess: false,
    },
  };

  const [fields, setFields] = useState(initialState.fields);
  const [alert, setAlert] = useState(initialState.alert);

  const handleAddRecipe = (event) => {
    event.preventDefault();

    setAlert({ message: "", isSuccess: false });

    axios
      .post("http://localhost:4000/api/v1/recipes", fields)
      .then(() =>
        setAlert({
          message: "Recipe Added",
          isSuccess: true,
        }),
      )
      .catch(() =>
        setAlert({
          message: "Server error. Please try again later.",
          isSuccess: false,
        }),
      );
  };

  const handleFieldChange = (event) => {
    const { name, value } = event.target;
    setFields({ ...fields, [name]: value });
  };

  const handleAddIngredient = () => {
    setFields((prevFields) => ({
      ...prevFields,
      ingredients: [
        ...prevFields.ingredients,
        { name: "", measurement: { value: "", metric: "" } },
      ],
    }));
  };

  const handleRemoveIngredient = (index) => {
    setFields((prevFields) => {
      const updatedIngredients = [...prevFields.ingredients];
      updatedIngredients.splice(index, 1);
      return { ...prevFields, ingredients: updatedIngredients };
    });
  };

  const handleIngredientChange = (index, field, value) => {
    const updatedIngredients = [...fields.ingredients];
    if (field === "measurement") {
      if (/^\d*\.?\d*$/.test(value) || value === "") {
        updatedIngredients[index].measurement.value = value;
      }
    } else if (field === "metric") {
      updatedIngredients[index].measurement.metric = value;
    } else {
      updatedIngredients[index][field] = value;
    }

    setFields({ ...fields, ingredients: updatedIngredients });
  };

  return (
    <div className="add-recipe">
      <form onSubmit={handleAddRecipe}>
        <div>
          <label htmlFor="title">
            Title:
            <input
              type="text"
              id="title"
              name="title"
              value={fields.title}
              onChange={handleFieldChange}
              maxLength={32}
              required
            />
          </label>
        </div>
        <div>
          <label htmlFor="cuisine">
            Cuisine:
            <select
              id="cuisine"
              name="cuisine"
              value={fields.cuisine}
              onChange={handleFieldChange}
              required
            >
              <option value="" disabled>
                --Select a cuisine--
              </option>
              <option value="Italian">Italian</option>
              <option value="Chinese">Chinese</option>
              <option value="Japanese">Japanese</option>
              <option value="Indian">Indian</option>
              <option value="Mexican">Mexican</option>
              <option value="French">French</option>
              <option value="Thai">Thai</option>
              <option value="Spanish">Spanish</option>
              <option value="Middle Eastern">Middle Eastern</option>
              <option value="Korean">Korean</option>
            </select>
          </label>
        </div>
        <div>
          <label htmlFor="dietaryRequirements">
            Dietary Requirements:
            <select
              id="dietaryRequirements"
              name="dietaryRequirements"
              value={fields.dietaryRequirements}
              onChange={handleFieldChange}
              required
            >
              <option value="" disabled>
                --Select a dietary requirement--
              </option>
              <option value="None">None</option>
              <option value="Vegetarian">Vegetarian</option>
              <option value="Vegan">Vegan</option>
              <option value="Gluten-Free">Gluten-Free</option>
              <option value="Dairy-Free">Dairy-Free</option>
              <option value="Pescatarian">Pescatarian</option>
            </select>
          </label>
        </div>
        <div>
          <label htmlFor="ingredients">
            Ingredients:
            {fields.ingredients.map((ingredient, index) => (
              <div key={ingredient} className="ingredient-field">
                <input
                  id={`ingredientName-${index}`}
                  type="text"
                  value={ingredient.name}
                  onChange={(e) =>
                    handleIngredientChange(index, "name", e.target.value)
                  }
                  required
                />
                <div className="measurement-metric-container">
                  <input
                    id={`measurement-${index}`}
                    placeholder="#"
                    type="number"
                    value={ingredient.measurement.value}
                    onChange={(e) =>
                      handleIngredientChange(
                        index,
                        "measurement",
                        e.target.value,
                      )
                    }
                    required
                    className="measurement-input"
                  />
                  <select
                    id={`metric-${index}`}
                    value={ingredient.measurement.metric}
                    onChange={(e) =>
                      handleIngredientChange(index, "metric", e.target.value)
                    }
                    required
                    className="metric-dropdown"
                  >
                    <option value="" disabled>
                      --Select a metric--
                    </option>
                    <option value="g">g</option>
                    <option value="kg">kg</option>
                    <option value="ml">ml</option>
                    <option value="l">l</option>
                    <option value="tbsp">tbsp</option>
                    <option value="tsp">tsp</option>
                  </select>
                </div>
                <div>
                  {index !== 0 && (
                    <button
                      className="form-button"
                      type="button"
                      onClick={() => handleRemoveIngredient(index)}
                    >
                      Remove
                    </button>
                  )}
                </div>
              </div>
            ))}
            <button
              className="form-button"
              type="button"
              onClick={handleAddIngredient}
            >
              Add Ingredient
            </button>
          </label>
        </div>

        <div>
          <label htmlFor="instructions">
            Instructions:
            <textarea
              id="instructions"
              name="instructions"
              value={fields.instructions}
              onChange={handleFieldChange}
              required
            />
          </label>
        </div>
        <div>
          <label htmlFor="prepTime">
            Prep Time (minutes):
            <input
              type="number"
              id="prepTime"
              name="prepTime"
              value={fields.prepTime}
              onChange={handleFieldChange}
              max={9999}
              required
            />
          </label>
        </div>
        <div>
          <label htmlFor="cookingTime">
            Cooking Time (minutes):
            <input
              type="number"
              id="cookingTime"
              name="cookingTime"
              value={fields.cookingTime}
              onChange={handleFieldChange}
              max={9999}
              required
            />
          </label>
        </div>
        <div>
          <label htmlFor="servings">
            Servings:
            <input
              type="number"
              id="servings"
              name="servings"
              value={fields.servings}
              onChange={handleFieldChange}
              max={9999}
              required
            />
          </label>
        </div>
        <div>
          <button className="form-button" type="submit">
            Add
          </button>
        </div>
      </form>
      <Alert message={alert.message} success={alert.isSuccess} />
    </div>
  );
};

export default AddRecipe;
