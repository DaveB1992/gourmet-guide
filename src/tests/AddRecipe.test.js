import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import AddRecipe from "../components/AddRecipe";
import axios from "axios";

describe("AddRecipe", () => {
  it("renders correctly", () => {
    const { asFragment } = render(<AddRecipe />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("updates form input correctly - Title", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const titleInput = getByLabelText("Title:");
    fireEvent.change(titleInput, { target: { value: "Pizza Margherita" } });

    expect(titleInput.value).toBe("Pizza Margherita");
  });

  it("updates form input correctly - Cuisine", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const cuisineSelect = getByLabelText("Cuisine:");
    fireEvent.change(cuisineSelect, { target: { value: "Italian" } });

    expect(cuisineSelect.value).toBe("Italian");
  });

  it("updates form input correctly - Dietary Requirements", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const dietaryRequirementsSelect = getByLabelText("Dietary Requirements:");
    fireEvent.change(dietaryRequirementsSelect, { target: { value: "None" } });

    expect(dietaryRequirementsSelect.value).toBe("None");
  });

  it("populates ingredient fields correctly", async () => {
    const { getByText, getAllByRole } = render(<AddRecipe />);

    const addIngredientButton = getByText("Add Ingredient");
    fireEvent.click(addIngredientButton);

    const ingredientNameInput = getAllByRole("textbox")[0];
    const measurementValueInput = getAllByRole("spinbutton")[0];
    const measurementMetricSelect = document.getElementById("metric-0");

    fireEvent.change(ingredientNameInput, { target: { value: "Tomatoes" } });
    fireEvent.change(measurementValueInput, { target: { value: "200" } });
    fireEvent.change(measurementMetricSelect, { target: { value: "g" } });

    await waitFor(() => {
      expect(ingredientNameInput.value).toBe("Tomatoes");
      expect(measurementValueInput.value).toBe("200");
      expect(measurementMetricSelect.value).toBe("g");
    });
  });

  it("updates form input correctly - Instructions", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const instructionsTextarea = getByLabelText("Instructions:");
    fireEvent.change(instructionsTextarea, {
      target: {
        value:
          "1. Preheat the oven to 220°C.\n2. Roll out the pizza dough.\n3. Spread tomato sauce on the dough.\n4. Add slices of mozzarella cheese and fresh basil leaves.\n5. Bake in the oven for 12-15 minutes or until the crust is golden and the cheese is bubbly and slightly browned.",
      },
    });

    expect(instructionsTextarea.value).toBe(
      "1. Preheat the oven to 220°C.\n2. Roll out the pizza dough.\n3. Spread tomato sauce on the dough.\n4. Add slices of mozzarella cheese and fresh basil leaves.\n5. Bake in the oven for 12-15 minutes or until the crust is golden and the cheese is bubbly and slightly browned."
    );
  });

  it("updates form input correctly - Prep Time", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const prepTimeInput = getByLabelText("Prep Time (minutes):");
    fireEvent.change(prepTimeInput, { target: { value: "20" } });

    expect(prepTimeInput.value).toBe("20");
  });

  it("updates form input correctly - Cooking Time", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const cookingTimeInput = getByLabelText("Cooking Time (minutes):");
    fireEvent.change(cookingTimeInput, { target: { value: "15" } });

    expect(cookingTimeInput.value).toBe("15");
  });

  it("updates form input correctly - Servings", () => {
    const { getByLabelText } = render(<AddRecipe />);
    const servingsInput = getByLabelText("Servings:");
    fireEvent.change(servingsInput, { target: { value: "2" } });

    expect(servingsInput.value).toBe("2");
  });
});

it("adds new ingredient field on button click", () => {
  const { getByText, getAllByRole } = render(<AddRecipe />);

  const addIngredientButton = getByText("Add Ingredient");
  fireEvent.click(addIngredientButton);

  const updatedIngredientFields = getAllByRole("textbox", { name: "" });
  expect(updatedIngredientFields.length).toBe(2);
});

it("removes ingredient field on button click", () => {
  const { getByText, queryByLabelText } = render(<AddRecipe />);

  const addIngredientButton = getByText("Add Ingredient");
  fireEvent.click(addIngredientButton);

  const removeButton = getByText("Remove");
  expect(removeButton).toBeInTheDocument();

  fireEvent.click(removeButton);

  const updatedIngredientField = queryByLabelText("Ingredient Name:");
  expect(updatedIngredientField).toBeNull();
});

it("submits the form and shows success alert", async () => {
  const { getByLabelText, getByText } = render(<AddRecipe />);

  const titleInput = getByLabelText("Title:");
  fireEvent.change(titleInput, { target: { value: "Pizza Margherita" } });
  const cuisineSelect = getByLabelText("Cuisine:");
  fireEvent.change(cuisineSelect, { target: { value: "Italian" } });
  const dietaryRequirementsSelect = getByLabelText("Dietary Requirements:");
  fireEvent.change(dietaryRequirementsSelect, { target: { value: "None" } });
  const instructionsTextarea = getByLabelText("Instructions:");
  fireEvent.change(instructionsTextarea, {
    target: {
      value:
        "1. Preheat the oven to 220°C.\n2. Roll out the pizza dough.\n3. Spread tomato sauce on the dough.\n4. Add slices of mozzarella cheese and fresh basil leaves.\n5. Bake in the oven for 12-15 minutes or until the crust is golden and the cheese is bubbly and slightly browned.",
    },
  });
  const prepTimeInput = getByLabelText("Prep Time (minutes):");
  fireEvent.change(prepTimeInput, { target: { value: "20" } });
  const cookingTimeInput = getByLabelText("Cooking Time (minutes):");
  fireEvent.change(cookingTimeInput, { target: { value: "15" } });
  const servingsInput = getByLabelText("Servings:");
  fireEvent.change(servingsInput, { target: { value: "2" } });

  const addButton = getByText("Add");
  jest.spyOn(axios, "post").mockResolvedValue({});
  fireEvent.click(addButton);

  await waitFor(() => {
    const successAlert = getByText("Recipe Added", {
      exact: false,
    });
    expect(successAlert).toBeInTheDocument();
  });
  axios.post.mockRestore();
});

it("shows server error alert when submission fails", async () => {
  const { getByText } = render(<AddRecipe />);

  jest.spyOn(axios, "post").mockRejectedValue(new Error("Server error"));

  const addButton = getByText("Add");
  fireEvent.click(addButton);

  await waitFor(() => {
    const errorAlert = getByText("Server error. Please try again later.", {
      exact: false,
    });
    expect(errorAlert).toBeInTheDocument();
  });

  axios.post.mockRestore();
});
