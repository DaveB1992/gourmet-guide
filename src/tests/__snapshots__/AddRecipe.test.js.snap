// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRecipe renders correctly 1`] = `
<DocumentFragment>
  <div
    class="add-recipe"
  >
    <form>
      <div>
        <label
          for="title"
        >
          Title:
          <input
            id="title"
            maxlength="32"
            name="title"
            required=""
            type="text"
            value=""
          />
        </label>
      </div>
      <div>
        <label
          for="cuisine"
        >
          Cuisine:
          <select
            id="cuisine"
            name="cuisine"
            required=""
          >
            <option
              disabled=""
              value=""
            >
              --Select a cuisine--
            </option>
            <option
              value="Italian"
            >
              Italian
            </option>
            <option
              value="Chinese"
            >
              Chinese
            </option>
            <option
              value="Japanese"
            >
              Japanese
            </option>
            <option
              value="Indian"
            >
              Indian
            </option>
            <option
              value="Mexican"
            >
              Mexican
            </option>
            <option
              value="French"
            >
              French
            </option>
            <option
              value="Thai"
            >
              Thai
            </option>
            <option
              value="Spanish"
            >
              Spanish
            </option>
            <option
              value="Middle Eastern"
            >
              Middle Eastern
            </option>
            <option
              value="Korean"
            >
              Korean
            </option>
          </select>
        </label>
      </div>
      <div>
        <label
          for="dietaryRequirements"
        >
          Dietary Requirements:
          <select
            id="dietaryRequirements"
            name="dietaryRequirements"
            required=""
          >
            <option
              disabled=""
              value=""
            >
              --Select a dietary requirement--
            </option>
            <option
              value="None"
            >
              None
            </option>
            <option
              value="Vegetarian"
            >
              Vegetarian
            </option>
            <option
              value="Vegan"
            >
              Vegan
            </option>
            <option
              value="Gluten-Free"
            >
              Gluten-Free
            </option>
            <option
              value="Dairy-Free"
            >
              Dairy-Free
            </option>
            <option
              value="Pescatarian"
            >
              Pescatarian
            </option>
          </select>
        </label>
      </div>
      <div>
        <label
          for="ingredients"
        >
          Ingredients:
          <div
            class="ingredient-field"
          >
            <input
              id="ingredientName-0"
              required=""
              type="text"
              value=""
            />
            <div
              class="measurement-metric-container"
            >
              <input
                class="measurement-input"
                id="measurement-0"
                placeholder="#"
                required=""
                type="number"
                value=""
              />
              <select
                class="metric-dropdown"
                id="metric-0"
                required=""
              >
                <option
                  disabled=""
                  value=""
                >
                  --Select a metric--
                </option>
                <option
                  value="g"
                >
                  g
                </option>
                <option
                  value="kg"
                >
                  kg
                </option>
                <option
                  value="ml"
                >
                  ml
                </option>
                <option
                  value="l"
                >
                  l
                </option>
                <option
                  value="tbsp"
                >
                  tbsp
                </option>
                <option
                  value="tsp"
                >
                  tsp
                </option>
              </select>
            </div>
            <div />
          </div>
          <button
            class="form-button"
            type="button"
          >
            Add Ingredient
          </button>
        </label>
      </div>
      <div>
        <label
          for="instructions"
        >
          Instructions:
          <textarea
            id="instructions"
            name="instructions"
            required=""
          />
        </label>
      </div>
      <div>
        <label
          for="prepTime"
        >
          Prep Time (minutes):
          <input
            id="prepTime"
            max="9999"
            name="prepTime"
            required=""
            type="number"
            value=""
          />
        </label>
      </div>
      <div>
        <label
          for="cookingTime"
        >
          Cooking Time (minutes):
          <input
            id="cookingTime"
            max="9999"
            name="cookingTime"
            required=""
            type="number"
            value=""
          />
        </label>
      </div>
      <div>
        <label
          for="servings"
        >
          Servings:
          <input
            id="servings"
            max="9999"
            name="servings"
            required=""
            type="number"
            value=""
          />
        </label>
      </div>
      <div>
        <button
          class="form-button"
          type="submit"
        >
          Add
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
